### PR TITLE
Add DrawGridTest

### PR DESCRIPTION
**Pull request by mtmccrea**
## Purpose and Motivation

This is a follow-on PR to the recently merged #5161, refactoring the `DrawGrid:*test` into  a new standalone class, removing clutter from `DrawGrid` and allowing access to the grid object under test as well as its enclosing view. This was brought up in the discussion of #5161, but deferred to its own PR.

There's also expanded and clarified the documentation, with examples and scdoc syntax formatting added.

## Types of changes

- Documentation
- Bug fix
- New feature
  - `DrawGrid:-numTicks` and `-tickSpacing` now allows setting x and y axes individually.

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

File: SCClassLibrary/Common/GUI/Base/Grid.sc [hunk 1]
```diff
@@ -119,19 +74,18 @@ DrawGrid {
 DrawGridX {
 
 	var <grid,<>range,<>bounds;
-	var <>font,<>fontColor,<>gridColor,<>labelOffset;
+	var <font,<fontColor,<gridColor,<labelOffset;
```
Comment (telephon) [inline SCClassLibrary/Common/GUI/Base/Grid.sc:77]: while you are at it, could you add spaces after the commas? Thanks!

**Review by telephon — CHANGES_REQUESTED** [review]
This is great, thank you so much for the detailed documentation. Just a few suggestions for better wording, and one interface issue to discuss.

File: HelpSource/Classes/DrawGrid.schelp [hunk 2]
```diff
@@ -117,54 +124,158 @@ returns:: Self.
...
+)
+::
+
+
+METHOD:: tickSpacing
+Set the emphasis::minimum:: spacing between grid lines ("ticks") for each axis. The number of grid lines will change with the view size, but won't be spaced less than this code::tickSpacing::, allowing you to control the density of grid lines. However if link::#-numTicks:: is not code::nil::, it takes precedence over code::tickSpacing::.
+
+See link::#Testing in a UserView#Examples:: below.
+
+argument:: xpx
```
[the remark this answers is not in the document]
Comment (mtmccrea) [reply HelpSource/Classes/DrawGrid.schelp:184]: I think the purpose is hinted at in the method name, `tickSpacing`, no? I think explicit units is helpful here, so it's not confused for, say, the units of the spec.

But if we do change it, should we then also change the arg for `DrawGridX:-tickSpacing` from `px` to `spacing`?

File: SCClassLibrary/Common/GUI/Base/Grid.sc (hunk 1 above)
Comment (mtmccrea) [reply SCClassLibrary/Common/GUI/Base/Grid.sc:77]: I actually rolled back a bunch of white space formatting changes in this PR, based on your [comment](https://github.com/supercollider/supercollider/pull/5161#issuecomment-1236568038) on in #5161. 

I was thinking of doing a separate PR that just overhauls all the style problems of this file at once. Sound OK?

File: HelpSource/Classes/AbstractGridLines.schelp [hunk 3]
```diff
@@ -35,100 +36,105 @@ CLASSMETHODS::
...
+code::
+(
+var g = GridLines([0, 10000].asSpec);
+"in val / rounded / not rounded".postln;
+10.collect{ 10000.rand }.sort.do({ |x|
+    postf("% / % / %\n", x, g.niceNum(x, true), g.niceNum(x, false))
+})
+)
+::
+The implementation is based on: http://books.google.de/books?id=fvA7zLEFWZgC&pg=PA61&lpg=PA61.
```
[the remark this answers is not in the document]
Comment (mtmccrea) [reply HelpSource/Classes/AbstractGridLines.schelp]: Thanks for pointing this out, I've updated the link and made the linked text a bibliographic entry so it's more informative at a glance.

File: HelpSource/Classes/AbstractGridLines.schelp [hunk 4]
```diff
@@ -1,32 +1,33 @@
...
-Subclasses of code::AbstractGridLines:: bind more tightly with the data they are representing. E.g. code::LinearGridLines:: represents data on linear scale, while code::ExponentialGridLines:: represents data on exponential scale. Future development work could add other subclasses like e.g. DegreeGridLines to be used to draw pitch degree GridLines behind a frequency plot.
-
-link::Classes/Spec:: has a .grid variable that points to its preferred code::AbstractGridLines:: subclass that should be used for graphing.
+code::AbstractGridLines:: shouldn't be instantiated directly. Instead, the link::Classes/GridLines:: factory class or the link::Classes/ControlSpec#-grid:: method should be used to return the appropriate code::AbstractGridLines:: subclass for the given spec.
 
 code::
 \freq.asSpec.grid
 ::
 
+Subclasses of code::AbstractGridLines:: bind more tightly with the data they represent. For example, code::LinearGridLines:: and code::ExponentialGridLines:: represent data on a linear and exponential scale, respectively.
```
[the remark this answers is not in the document]
Comment (mtmccrea) [reply HelpSource/Classes/AbstractGridLines.schelp]: That language was in the original `AbstractGridLines` doc, so I can't speak for what was meant by that, but I agree with your comment and have modified it.

File: HelpSource/Classes/AbstractGridLines.schelp [hunk 5]
```diff
@@ -35,100 +36,105 @@ CLASSMETHODS::
 METHOD:: new
 
 argument:: spec
-The ControlSpec that defines the minimum and maximum values, warp and step.
+A link::Classes/ControlSpec:: that defines the numerical range, warp, and step.
 
-returns:: a AbstractGridLines
+returns:: The newly instantiated object.
```
[the remark this answers is not in the document]
Comment (mtmccrea) [reply HelpSource/Classes/AbstractGridLines.schelp]: Because all of the method descriptions of `AbstractGridLines` are inherited by its subclasses via `copymethod::`, I was trying to make it generic enough to apply to the subclasses as well.
I've changed it to "An instance of this class."

But if it's preferred to state the class, I can just add "An instance of xxx." in each of the subclasses.

File: HelpSource/Classes/AbstractGridLines.schelp [hunk 6]
```diff
@@ -35,100 +36,105 @@ CLASSMETHODS::
...
-returns:: self
+discussion::
+code::nil.asGrid:: returns a code::BlankGridLines:: (a subclass of code::AbstractGridLines::) which produces no drawn lines. So if code::nil:: is passed to the strong::grid:: argument of code::DrawGrid::, no lines are drawn on the corresponding axis.
 
-METHOD:: niceNum
-Based on: http://books.google.de/books?id=fvA7zLEFWZgC&pg=PA61&lpg=PA61
 
-This rounds a value to a logical nice number.  It is mostly used to support internal calculation, though it may be useful for other applications.
+METHOD:: niceNum
+Rounds a value to a logical "nice number".  This method is used to support internal calculations for determining grid line values, though it may be useful for other applications.
```
[the remark this answers is not in the document]
Comment (mtmccrea) [reply HelpSource/Classes/AbstractGridLines.schelp]: Changed to: "Rounds a value to a logical "nice number" that is within the same order of magnitude.", and moved the comment on its use to the `discussion::`.

**Comment by mtmccrea** [comment]
Thanks @telephon for jumping on this so quickly! I've responded to all of your comments, and there are a couple open questions in response.

**Comment by mtmccrea** [comment]
I'll also ping @dyfer, since he was a primary author of the PR this follows on.

File: SCClassLibrary/Common/GUI/Base/Grid.sc [hunk 7]
```diff
@@ -106,6 +53,14 @@ DrawGrid {
...
 		y.grid = g;
 	}
+	tickSpacing_ { arg xpx, ypx;
+		x.tickSpacing = xpx;
+		y.tickSpacing = ypx;
+	}
+	numTicks_ { arg numx, numy;
+		x.numTicks = numx;
+		y.numTicks = numy;
+	}
```
[the remark this answers is not in the document]
Comment (mtmccrea) [reply SCClassLibrary/Common/GUI/Base/Grid.sc:63]: The main motivation was that having the convenience method `tickSpacing_` that sets both axes to the _same_ spacing doesn't make much sense considering the scales of the data is usually different on each axis, as well as the aspect ratio. Same for `numTicks_`.

Another advantage is not having to think about individual axes as separate objects.

File: HelpSource/Classes/AbstractGridLines.schelp (hunk 6 above)
[the remark this answers is not in the document]
Comment (telephon) [reply HelpSource/Classes/AbstractGridLines.schelp]: good solution.

File: HelpSource/Classes/DrawGrid.schelp (hunk 2 above)
[the remark this answers is not in the document]
Comment (telephon) [reply HelpSource/Classes/DrawGrid.schelp:184]: ah yes, I had overlooked this. Then I would say, name it to `h` and `v`, or `x` and `y`. We never indicate `px` in the GUI classes (see e.g. `Rect`).

File: SCClassLibrary/Common/GUI/Base/Grid.sc (hunk 1 above)
Comment (telephon) [reply SCClassLibrary/Common/GUI/Base/Grid.sc:77]: OK!

File: SCClassLibrary/Common/GUI/Base/Grid.sc (hunk 7 above)
[the remark this answers is not in the document]
Comment (telephon) [reply SCClassLibrary/Common/GUI/Base/Grid.sc:63]: I agree, because this would show in the code _that_ you can access these objects separately.

File: HelpSource/Classes/DrawGrid.schelp (hunk 2 above)
[the remark this answers is not in the document]
Comment (telephon) [reply HelpSource/Classes/DrawGrid.schelp:184]: (but see comment below)

[the remark this answers is not in the document]
Comment (mtmccrea) [reply HelpSource/Classes/DrawGrid.schelp]: OK I've changed the args to `x` and `y` in both the `-tickSpacing` and `-numTicks` methods.
>(but see comment below)

I'm not sure which comment this refers to...

File: SCClassLibrary/Common/GUI/Base/Grid.sc [hunk 8]
```diff
@@ -483,13 +464,87 @@ ExponentialGridLines : AbstractGridLines {
...
 BlankGridLines : AbstractGridLines {
 
-	getParams {
-		^()
+	getParams { ^() }
+
+	prCheckWarp { }
+}
+
+DrawGridTest : DrawGrid {
```
[the remark this answers is not in the document]
Comment (telephon) [reply SCClassLibrary/Common/GUI/Base/Grid.sc:474]: Or is it something that could be of a more general purpose, then it would be a `DrawGridView` and may be a subclass of `View`.

[the remark this answers is not in the document]
Comment (dyfer) [reply SCClassLibrary/Common/GUI/Base/Grid.sc:474]: Well, the original implementation was `DrawGrid.test` so the class name is related to that, AFAIU.
However, I do get that it can be confused with unit tests so maybe another name would be better.

File: SCClassLibrary/Common/GUI/Base/Grid.sc (hunk 7 above)
[the remark this answers is not in the document]
Comment (dyfer) [reply SCClassLibrary/Common/GUI/Base/Grid.sc]: > The main motivation was that having the convenience method tickSpacing_ that sets both axes to the same spacing doesn't make much sense

@mtmccrea yes, I get that. But my point was do we really need that convenience method at all, if we can do tickSpacing on individual grids... which you address in the second paragraph.

@telephon I'm not sure whether you agree with adding the convenience method or not adding it? :)

File: SCClassLibrary/Common/GUI/Base/Grid.sc (hunk 8 above)
[the remark this answers is not in the document]
Comment (telephon) [reply SCClassLibrary/Common/GUI/Base/Grid.sc:474]: Yes, "view" would only be appropriate if it were generalized to something multi-purpose. I like `DrawGridDisplay`, and the corresponding verb would work, too.

File: SCClassLibrary/Common/GUI/Base/Grid.sc (hunk 7 above)
[the remark this answers is not in the document]
Comment (mtmccrea) [reply SCClassLibrary/Common/GUI/Base/Grid.sc]: Thanks for considering this, @dyfer, 
>do we really need that convenience method at all, if we can do tickSpacing on individual grids... which you address in the second paragraph.

In addition to those reasons I gave, another reason is that there are similar convenience methods for multiple other properties of `DrawGridX/Y` (bounds, font, fontColor, gridColors). So in the interest of a complete implementation, again which models the behavior of `Plotter`, I'd say yes.

File: SCClassLibrary/Common/GUI/Base/Grid.sc (hunk 8 above)
[the remark this answers is not in the document]
Comment (mtmccrea) [reply SCClassLibrary/Common/GUI/Base/Grid.sc:474]: This brings up a conceptual question about the need for the class at all, which I'll bring up in the main conversation thread.

I like "Display" as well. The method `*test` was in the original implementation. It refers to its intended use, but if we're ok with a breaking change, we could consider changing it.

**Comment by mtmccrea** [comment]
Thanks @dyfer and @telephon for your comments above about naming the class. It brings up a conceptual point as to whether this class is even necessary. `DrawGridTest` is introduced here to separate code involved with creating a window for "testing" the `DrawGrid`—functionality which was built in to the original implementation with the `*test` method.

>making it a subclass of DrawGrid looks a bit funny, because its main purpose seems to be to be a display.

At some point along the way, a workflow was introduced of altering the grid's properties _after_ creating this mockup display. This created more requirements for storing the state of the grid and view in order to handle refreshing when the grid was updated (hence being a subclass of `DrawGrid`). A proper way to do this is with a model-view-controller design, but that isn't how it was designed and that's overkill to add.

I would suggest that if we revise this intended "test" workflow to only support "snapshot" previewing of the current state. It eliminates the need for the class and the test method could be restored to `DrawGrid`. The intended workflow would be (and documented as):
- create a `DrawGrid`
- modify its properties
- call `.test` to create a window that previews it (or `.display` if we're ok with the breaking change of changing the method)
- close the preview window
- if you want to make changes, do so and call `.test` again, etc.

In other words, the snapshot is static and doesn't update if the `DrawGrid` properties are updated. You need to `.test` it again if you change it.

What do you think?

File: SCClassLibrary/Common/GUI/Base/Grid.sc (hunk 7 above)
[the remark this answers is not in the document]
Comment (telephon) [reply SCClassLibrary/Common/GUI/Base/Grid.sc]: ok, well then that is a good reason to have them. It may not have been a good idea (by me) to add these to `Plotter` in the first place, but well :) …

[the remark this answers is not in the document]
Comment (mtmccrea) [reply SCClassLibrary/Common/GUI/Base/Grid.sc]: Actually I appreciate being able to set the properties of all subplots from `Plotter` with arrays :)

[the remark this answers is not in the document]
Comment (telephon) [reply SCClassLibrary/Common/GUI/Base/Grid.sc]: :)

File: SCClassLibrary/Common/GUI/Base/Grid.sc (hunk 8 above)
[the remark this answers is not in the document]
Comment (telephon) [reply SCClassLibrary/Common/GUI/Base/Grid.sc:474]: I think it would be ok to have a breaking chnage here, because it is nothing that should be used in systems, but more in exploration.

**Comment by telephon** [comment]
Very good, I think!

**Comment by telephon** [comment]
It is already possible to live update and tweak, with an example like this:
```supercollider
(
w = Window.new.front;
u = UserView(w,w.bounds.size.asRect);
​
// the Spec can define its preferred grid system
x = \lofreq.asSpec.grid;
y = \amp.asSpec.grid;
d = DrawGrid(u.bounds.size.asRect.insetBy(20), x, y);
​
u.drawFunc = { d.draw };
u.resize_(5).onResize = { |u| d.bounds = u.bounds.size.asRect.insetBy(20) }
)


d.horzGrid = ExponentialGridLines(\freq); // change on the fly

```

**Comment by mtmccrea** [comment]
I should have been more specific. The refreshing issue is (mostly) in the case of instantiation with `*test`. Though there are some caveats on refreshing those parameters which rely on new `commands` being constructed, which require not only a `-refresh` but also `clearCache` (this is fixed in this PR as well - [f529d9f](https://github.com/supercollider/supercollider/pull/5895/commits/f529d9f980eff9196402c1da214e1eb77c6ca663)). 
```supercollider
(
w = Window.new.front;
u = UserView(w,w.bounds.size.asRect);
​
// the Spec can define its preferred grid system
x = \lofreq.asSpec.grid;
y = \amp.asSpec.grid;
d = DrawGrid(u.bounds.size.asRect.insetBy(20), x, y);
​
u.drawFunc = { d.draw };
u.resize_(5).onResize = { |u| d.bounds = u.bounds.size.asRect.insetBy(20) }
)

d.gridColors_([Color.red, Color.blue]); u.refresh // only one axis color updates?
d.gridColors_([Color.red, Color.blue]); d.clearCache; u.refresh // now works

// Instantiate using *test
t = DrawGrid.test(x, y, (500@300).asRect);
t.gridColors_([Color.red, Color.blue]); // no access to the view to refresh :(
```

>In other words, the snapshot is static and doesn't update if the DrawGrid properties are updated. You need to .test it again if you change it.
What do you think?
>>Very good, I think!

OK cool, I'll have a look at revising it.

**Comment by telephon** [comment]
> > In other words, the snapshot is static and doesn't update if the DrawGrid properties are updated. You need to .test it again if you change it.
> > What do you think?
> > > Very good, I think!
> 
> OK cool, I'll have a look at revising it.

As you suggested, we could also leave it out, if that seems reasonable.

**Comment by mtmccrea** [comment]
OK @telephon and @dyfer, I've pushed changes which hopefully simplify things.

I decided on the method `-preview` for this use case, and because we're OK with breaking changes on this use case, I _removed_ the `*test` class method as well. I don't think the workflow benefits from it more than an instance method, and it frees us from having to decide whether it should return the `DrawGrid` instance or the `UserView` created internally.

The workflow now looks like this:
```supercollider
(
x = [0, 10].asSpec.units_("sec");
y = [0, 1].asSpec.units_("amp");
d = DrawGrid((500@250).asRect, x.grid, y.grid);
// set its properties before testing
d.x.tickSpacing_(25);
d.linePattern_(FloatArray[10.0, 5.0, 2.0, 5.0]);
// generate a preview
~testView = d.preview;
)

// Use DrawGrid's convenience methods to set underlying
// grid properties, then refresh UserView.
d.tickSpacing_(50, 25).linePattern_(FloatArray[1.0]);
~testView.refresh;

// or just call .preview again and it will refresh the existing view
d.linePattern_(FloatArray[10.0, 5.0, 2.0, 5.0]).preview;
```
Note that the internal `UserView` (the `uview` instance variable) has no getters or setters—it's considered "private" in the sense that it's managed internally (cleaned up when the window is closed). You can still access the view though—just call `-preview` again, which will return the view if the window is still open (refreshing it also, so no need to call `refresh`) or otherwise create a new window/view and return it.

If this looks good, I'll update the help docs as well.

File: SCClassLibrary/Common/GUI/Base/Grid.sc [hunk 9]
```diff
@@ -44,8 +44,9 @@ DrawGrid {
 		y.fontColor = c;
 	}
 	gridColors_ { arg colors;
-		x.gridColor = colors[0];
-		y.gridColor = colors[1];
+		colors = colors.as(Array);
+		x.gridColor = colors.wrapAt(0);
```
Comment (telephon) [inline SCClassLibrary/Common/GUI/Base/Grid.sc:42]: why `wrapAt`here? Is it to make it forward compatible with an n-dimensional grid?

**Comment by telephon** [comment]
> I decided on the method `-preview` for this use case, and because we're OK with breaking changes on this use case, I _removed_ the `*test` class method as well.

good, that makes sense.

Normally, I'd tend to avoid an internal view variable, but well, in this case I suppose it is ok because the class is a gui class anyhow.

File: SCClassLibrary/Common/GUI/Base/Grid.sc (hunk 9 above)
Comment (mtmccrea) [reply SCClassLibrary/Common/GUI/Base/Grid.sc:42]: In case only a single color is passed in, both axes get the same color.

Comment (telephon) [reply SCClassLibrary/Common/GUI/Base/Grid.sc:42]: hm, ok, seems convenient, although in `gridColors = Color.black` the plural s is a little tilted. What about just write `gridColors = Color.black.dup` ?

**Comment by mtmccrea** [comment]
> Normally, I'd tend to avoid an internal view variable, but well, in this case I suppose it is ok because the class is a gui class anyhow.

I agree, the suggestion of an internal view in the previous PR was what started this detour to creating a dedicated "Test" class. But now that we've seen there's compromises with both approaches, I think this is a pretty lightweight, simplified solution.

So if this approach looks ok I'll go ahead and update the accompanying docs?

File: SCClassLibrary/Common/GUI/Base/Grid.sc (hunk 9 above)
Comment (mtmccrea) [reply SCClassLibrary/Common/GUI/Base/Grid.sc:42]: With this implementation, both `gridColors = Color.black` and `gridColors = Color.black.dup` would both work, which seems desirable from a user perspective.
Is the concern that the intention in the method's source is unclear?

**Review by telephon — APPROVED** [review]

**Comment by dyfer** [comment]
Thanks @mtmccrea and @telephon for your work on this.

I just wanted to double check - do we really want to remove `*test` method? While it's unlikely used "in production", ppl might have code for... well, trying out ("testing") grids. Could it be deprecated instead?

**Comment by telephon** [comment]
I think it could be fully deprecated, i.e. not functional, but with a suggestion what to do.

**Comment by dyfer** [comment]
@mtm could you add the suggested deprecation in https://github.com/supercollider/supercollider/tree/develop/SCClassLibrary/deprecated/3.13/GUI ?
(I wouldn't mind a functional deprecation, i.e. the one that still gives the expected result, but a non-functional one that just displayed a warning would be fine as well in this case I think)

**Comment by telephon** [comment]
Looks good to me.
